### PR TITLE
Various fixes

### PR DESCRIPTION
**Pull request by franz**
some bugfixes & documentation updates

### CODE REVIEW

File: src/CHIPDriver.cc [hunk 1]
```diff
@@ -131,8 +140,9 @@ extern hipError_t CHIPReinitialize(const uintptr_t *NativeHandles,
   // TODO Check configuration for what backends are configured
   if (!CHIPBackendType.compare("opencl")) {
     if (UsingDefaultBackend)
-      logWarn("CHIP_BE was not set. Defaulting to OPENCL");
-    logDebug("CHIPBE=OPENCL... Initializing OpenCL Backend");
+      logDebug("CHIP_BE was not set. Defaulting to OPENCL");
```
Comment (pvelesko) [inline src/CHIPDriver.cc:143]: probably want to default to L0 now?

File: src/backend/Level0/CHIPBackendLevel0.cc [hunk 2]
```diff
@@ -1448,21 +1448,21 @@ void *CHIPContextLevel0::allocateImpl(size_t Size, size_t Alignment,
...
-      ZeDeviceProps_() {
-  ZeDeviceProps_.pNext = nullptr;
-  assert(Ctx_ != nullptr);
-}
-CHIPDeviceLevel0::CHIPDeviceLevel0(ze_device_handle_t &&ZeDev,
+CHIPDeviceLevel0::CHIPDeviceLevel0(ze_device_handle_t ZeDev,
                                    CHIPContextLevel0 *ChipCtx, int Idx)
     : CHIPDevice(ChipCtx, Idx), ZeDev_(ZeDev), ZeCtx_(ChipCtx->get()),
       ZeDeviceProps_() {
-  ZeDeviceProps_.pNext = nullptr;
```
Comment (pvelesko) [inline src/backend/Level0/CHIPBackendLevel0.cc:1462]: This needs to remain in place due to how clang initialises structure members.

File: src/logging.cc [hunk 3]
```diff
@@ -8,10 +8,9 @@ void setupSpdlog() { std::call_once(SpdlogWasSetup, &_setupSpdlog); }
 
 void _setupSpdlog() {
   spdlog::set_default_logger(spdlog::stderr_color_mt("CHIP"));
-  spdlog::set_level(spdlog::level::debug);
   spdlog::set_pattern("%n %^%l%$ [TID %t] [%E.%F] : %v");
 
-  spdlog::level::level_enum SpdLogLevel;
```
Comment (pvelesko) [inline src/logging.cc:14]: We should probably set the default level to "debug" if compiled in Debug and to "warn" for release ?

**Review by pvelesko — CHANGES_REQUESTED** [review]
a few small things

File: src/CHIPDriver.cc (hunk 1 above)
Comment (franz) [reply src/CHIPDriver.cc:143]: I'm not sure. Moved discussion to #93

File: src/backend/Level0/CHIPBackendLevel0.cc (hunk 2 above)
Comment (franz) [reply src/backend/Level0/CHIPBackendLevel0.cc:1462]: ok, fixed

File: src/logging.cc (hunk 3 above)
Comment (franz) [reply src/logging.cc:14]: fixed

**Comment by pvelesko** [comment]
Please re-request the the review once you are done @franz

File: CMakeLists.txt [hunk 4]
```diff
@@ -156,7 +162,7 @@ endif()
 string(TOUPPER "${CMAKE_BUILD_TYPE}" uppercase_CMAKE_BUILD_TYPE)
 if(uppercase_CMAKE_BUILD_TYPE STREQUAL "DEBUG")
   add_compile_options("-Wall")
-  list(APPEND CHIP_SPV_DEFINITIONS SPDLOG_ACTIVE_LEVEL=SPDLOG_LEVEL_TRACE)
+  list(APPEND CHIP_SPV_DEFINITIONS SPDLOG_ACTIVE_LEVEL=SPDLOG_LEVEL_DEBUG)
```
[the remark this answers is not in the document]
Comment (pvelesko) [reply CMakeLists.txt:165]: Looking at `logging.cc` seems like this should work fine

[the remark this answers is not in the document]
Comment (pvelesko) [reply CMakeLists.txt:165]: just compiled using `-DCMAKE_BUILD_TYPE=RelWithDebInfo`, it showed me the expected warnings but I was unable to get the trace info after setting `export CHIP_LOGLEVEL=trace`

**Review by pvelesko — CHANGES_REQUESTED** [review]
comments

File: CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (franz) [reply CMakeLists.txt]: Right, setting SPDLOG_ACTIVE_LEVEL disables logging below that level completely. Fixed back to the original CMake code.

WRT trace not working with RelWithDebInfo - that is expected; the CMake code currently disables all messages below "warn" level on all build types other than Debug. TBH we could just make all logging available to all builds. I don't think it has a huge impact on performance.

File: README.md [hunk 5]
```diff
@@ -105,7 +105,7 @@ If you do not provide this value, `hipcc` will check for existance of the follow
 /opt/rocm
 ```
 
-### Compiling CUDA application directly with CHIP-SPV
+## Compiling CUDA application directly with CHIP-SPV
 
 Compilation of CUDA sources without changing the sources, can be done in two ways. The first is to replace calls of the nvcc compiler with calls of the wrapper script <CHIP-install-path>/bin/cuspv in Makefiles. The wrapper script will call clang with the correct flags. The other way is possible when using CMake: use `find_package(HIP REQUIRED CONFIG)` and then use `target_link_libraries(<YOUR_TARGET> hip::CHIP hip::host hip::device)`. However the project must be compiled with Clang (a version supported by HIP). Note that it's not necessary to have Nvidia's CUDA installed.
```
[the remark this answers is not in the document]
Comment (franz) [reply README.md:110]: fixed

File: Development.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+## Runtime
+
+The runtime implements the HIP API. It consists mainly of these files:
+
+* `src/CHIPBindings.cc` - implementation of HIP API functions like e.g. `hipEventRecord` by using the abstract backend classes.
+* `src/CHIPBackend.{cc,hh}` - a set of abstract backend classes used by the HIP API functions.
+* `src/backend/Level0/CHIPBackendLevel0.{cc,hh}` - Level0 implementation of the abstract backend.
+* `src/backend/OpenCL/CHIPBackendOpenCL.{cc,hh}` - OpenCL implementation of the abstract backend.
+* `src/Driver.{cc,hh}` - implementation of (un)initialization routines.
```
[the remark this answers is not in the document]
Comment (franz) [reply Development.md:99]: fixed

File: CMakeLists.txt [hunk 7]
```diff
@@ -217,10 +219,7 @@ add_subdirectory(./samples samples)
 add_dependencies(CHIP devicelib_bc LLVMHipPasses)
 add_dependencies(samples CHIP)
 
-# CHIP binaries
 add_subdirectory(bin)
-
-# CHIP-SPV CMAKE DEPENDENCIES
```
[the remark this answers is not in the document]
Comment (franz) [reply CMakeLists.txt:223]: fixed

File: CMakeLists.txt [hunk 8]
```diff
@@ -184,24 +201,9 @@ message(STATUS "CHIP-SPV will be installed to: ${CMAKE_INSTALL_PREFIX}")
...
-  find_library(LevelZero_LIBRARY NAMES ze_loader PATHS ENV LD_LIBRARY_PATH ./ REQUIRED)
-  message(STATUS "LevelZero_LIBRARY: ${LevelZero_LIBRARY}")
-endif()
 
-set(CMAKE_THREAD_PREFER_PTHREAD TRUE)
-set(THREADS_PREFER_PTHREAD_FLAG TRUE)
-find_package(Threads REQUIRED)
-set(PTHREAD_LIBRARY Threads::Threads)
+###############################################################################
+# CHIP binaries & tests
```
[the remark this answers is not in the document]
Comment (franz) [reply CMakeLists.txt:206]: fixed

File: CMakeLists.txt [hunk 9]
```diff
@@ -27,15 +54,21 @@ set(CHIP_SRC
...
+    src/backend/OpenCL/SVMemoryRegion.cc)
+endif()
+
+if(LevelZero_LIBRARY)
+  list(APPEND CHIP_SRC
+    src/backend/Level0/CHIPBackendLevel0.cc)
+endif()
+
 set(CHIP_SRC_DIR ${CMAKE_CURRENT_SOURCE_DIR} CACHE INTERNAL "CHIP-SPV source directory")
-set(CLANG_VERSION_LESS_14 FALSE CACHE INTERNAL "Clang version < 14.0")
```
[the remark this answers is not in the document]
Comment (franz) [reply CMakeLists.txt:37]: moved the related code into FindLLVM.cmake, but i can add them back there - however, is there a good reason to cache a version comparison ?

File: README.md [hunk 10]
```diff
@@ -119,6 +119,92 @@ You can find various HIP applications here: <https://github.com/CHIP-SPV/hip-tes
...
+
+there are also two APIs for asynchronous interoperability:
+
+* hipGetNativeEventFromHipEvent takes hipEvent_t and returns an OpenCL/Level0 event handle as void*
+* hipGetHipEventFromNativeEvent takes OpenCL/Level0 event handle, and returns a hipEvent_t as void*
+
+Before using the hipGetNativeEvent, the event must be recorded in a Stream.
+
+With OpenCL, both get*Event APIs increase the refcount of the cl_event and each side (HIP and the user application) are responsible for releasing the event when they’re done with it.
+With Level0, the event pool from which it was allocated remains the responsibility of the side that allocated it (e.g. getNativeEvent returns a ze_event handle but the pool is still managed and released by CHIP-SPV). This could lead to issues if e.g. a pool is released but an event allocated from it still exists as a dependency in some command enqueued by the opposite API. This can be resolved by keeping all eventpools allocated throughout application runtime.
```
[the remark this answers is not in the document]
Comment (franz) [reply README.md]: fixed the wording

**Comment by pvelesko** [comment]
All the comments I left were resolved so I am merging this branch.

**Pull request merged**